### PR TITLE
Granular scaffolder permissions

### PR DESCRIPTION
**Pull request by Zaperex**
## Hey, I just made a Pull Request!

Converted the `scaffolder.task.read` and `scaffolder.task.cancel` permissions into resource permissions, and added a rule to match for tasks owned by a specific user to allow conditional policies that restrict access to tasks to only their owners.

Also added a `scaffolder.task.admin.read` permission to grant access to all tasks (mainly for the `GET /v2/tasks` endpoint), and it will take precedence over the `scaffolder.task.read` permission policies if defined.

Fixes https://github.com/backstage/backstage/issues/25616

[Permission Policies.webm](https://github.com/user-attachments/assets/a2e232c0-4f6f-4dd9-bbfe-598cdf8172df)

Used the following custom permission policy in the recording above:
```ts
export class CustomPermissionPolicy implements PermissionPolicy {
  async handle(
    request: PolicyQuery,
    user?: PolicyQueryUser,
  ): Promise<PolicyDecision> {

    if (isPermission(request.permission, taskReadPermission)) {
      if (user?.info.userEntityRef === 'user:default/zaperex') {
        return {
          result: AuthorizeResult.ALLOW,
        };
      }
      if (user?.info.userEntityRef === 'user:default/datskrub') {
        return {
          result: AuthorizeResult.DENY,
        };
      }
      return createScaffolderTaskConditionalDecision(
        request.permission,
        scaffolderTaskConditions.isTaskOwner({
          createdBy: user?.info.userEntityRef ?? '',
        })
      )
    }
    return { result: AuthorizeResult.ALLOW };
  }
}
```
#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by Zaperex** [comment]
Do we want to completely hide the `Task List` field if users don't have permissions, or is just showing the 403 error banner sufficient when they click into it?

**Comment by benjdlambert** [comment]
Hey @Zaperex, not sure that we need the `admin` permission, I wonder if it's better just to push that to the policy instead and make the decision there instead.

**Comment by Zaperex** [comment]
sure, I'll make the change

**Review by camilaibs — APPROVED** [review]
It looks good to me, thanks for opening the pull request, but I would appreciate a review from @backstage/scaffolder-maintainers.

**Review by acierto — APPROVED** [review]

**Comment by benjdlambert** [comment]
So on the whole I think that this is OK, although after chatting with some folks internally I think that there's a chance that this could be a breaking change, but only for the `alpha` exports as this is permissions, which marked as `alpha` for these exact reasons.

Consider the following policy:

```ts
export class AllowAllPermissionPolicy implements PermissionPolicy {
  async handle(
    request: PolicyQuery,
  ): Promise<PolicyDecision> {
    if (isResourcePermission(request.permission)) {
      // The only resource permissions in my Backstage instance
      // are catalog permissions, so we can just return a catalog
      // decision here.
      return createCatalogConditionalDecision(
        request.permission,
        catalogConditions.isEntityOwner({
          claims: user?.info.ownershipEntityRefs ?? [],
        }),
      );
    } 

    return {
      result: AuthorizeResult.ALLOW,
    };
  }
}
```

This policy would do the right thing before and the wrong thing after. I think we need to call out some of this as breaking in the changeset to the `/alpha` exports in particular.

It does bring about the question that is this possible to ship in a non-breaking way, and I don't think it's possible with the permissions framework right now, i think we might want to take a look at scenarios like this ahead of moving permissions to v1.

**Review by benjdlambert — CHANGES_REQUESTED** [review]
As above ^ :pray:

**Comment by benjdlambert** [comment]
@Zaperex probably gonna merge this after the release goes out tomorrow as I feel it's a little bit risky just to ship this out in a mainline release tomorrow, wanna do some testing with it first before if that's OK? :pray:

**Comment by Zaperex** [comment]
Sure no problem

**Review by mounicarajput — APPROVED** [review]

**Comment by benjdlambert** [comment]
@Zaperex small conflict crept in here :pray:

File: plugins/scaffolder-backend/src/util/checkPermissions.ts
Comment (vinzscam) [inline plugins/scaffolder-backend/src/util/checkPermissions.ts:49]: is there any case where `throwError: false` is passed? I couldn't find it

File: plugins/scaffolder-backend/src/service/router.ts
Comment (vinzscam) [inline plugins/scaffolder-backend/src/service/router.ts:168]: this should be

```suggestion
> = PermissionRule<SerializedTask, {}, typeof RESOURCE_TYPE_SCAFFOLDER_TASK, TParams>;
```

or 

```suggestion
> = PermissionRule<TaskSpec, {}, typeof RESOURCE_TYPE_SCAFFOLDER_TASK, TParams>;
```

depending on what @backstage/scaffolder-maintainers recommend.

File: plugins/scaffolder-backend/src/service/rules.ts
Comment (vinzscam) [inline plugins/scaffolder-backend/src/service/rules.ts:146]: following the `PermissionRule`'s suggestion above, this would be `task` an the logic of the function should be adjusted accordingly. 

```suggestion
  apply: (task, { createdBy }) => {
```

File: plugins/scaffolder-backend/src/service/router.ts
Comment (vinzscam) [inline plugins/scaffolder-backend/src/service/router.ts:638]: mmm there is no task at this point 🤔

I'm not sure how this logic works, but ideally we would need to filter out only the tasks that the user shouldn't be able to see

**Review by vinzscam — CHANGES_REQUESTED** [review]

File: plugins/scaffolder-backend/src/service/router.ts
Comment (Zaperex) [reply plugins/scaffolder-backend/src/service/router.ts:168]: Ah so I chose to use `string` instead since this rule was to check for user ownership. Since we had a filter in the `/tasks` request to grab all tasks owned by a specific user, I thought it would be more performant to just check if the user is allowed to use the specified `userEntityRef` for the endpoint filter to avoid an unnecessary database query.

File: plugins/scaffolder-backend/src/service/router.ts
Comment (Zaperex) [reply plugins/scaffolder-backend/src/service/router.ts:638]: Ah so the idea as explained above is that we're checking whether the `createdBy` filter specified by the user is allowed before make the database query. If they're allowed to use the `userEntityRef` specified in the `createdBy` query, then we could just use the built-in `createdBy` filter to restrict the list of tasks the user is able to see.

File: plugins/scaffolder-backend/src/util/checkPermissions.ts
Comment (Zaperex) [reply plugins/scaffolder-backend/src/util/checkPermissions.ts:49]: Ah that might have been some leftover code from when I had an explicit admin permission. Let me revert this back to what it was before.

File: plugins/scaffolder-backend/src/service/rules.ts
Comment (Zaperex) [reply plugins/scaffolder-backend/src/service/rules.ts:146]: As discussed above, I made it a string as to match on the filter so that the check is made before the database query is made. Then if the specified user is allowed to user the filter, they then use the built-in `createdBy` filter to grab the users.

Comment (Zaperex) [reply plugins/scaffolder-backend/src/service/rules.ts:146]: I noticed that recently a new change was added to the `/v2/tasks` endpoint where it now accepts filters that might include a list of `createdBy` `userEntityRef`s. Would this be for filtering for tasks owned by a list of users? If so, I can update my PR for that case as well since the original purpose of this PR was restrict users to only tasks they own.

**Comment by benjdlambert** [comment]
@vinzscam can you take another look at this when you get a sec :pray:

**Comment by benjdlambert** [comment]
See comment here https://github.com/backstage/backstage/pull/26946#issuecomment-2396024126 also for some additional work that I'd like to resolve before merging this.

File: plugins/scaffolder-backend/src/service/router.ts
Comment (vinzscam) [reply plugins/scaffolder-backend/src/service/router.ts:165]: yeah this doesn't mean that we have to query the database again, the filtering logic is fine! 
But if we define `string` instead of `TaskSpec` here, we might have troubles in the future in case we introduce more rules. Let's say we'll introduce `hasMetadata` rule, we won't be able to access the entire `TaskSpec` but only the `owner` as string.

File: plugins/scaffolder-backend/src/service/router.ts
Comment (vinzscam) [reply plugins/scaffolder-backend/src/service/router.ts:635]: I see! This works right now, but it won't work in the future if we introduce more task rules that needs the task data.
This makes things a bit tricky in case `createdBy` contains more than one item, as highlighted in your TODO.
Here we have two choices, either we transform the rule in a db query, or we authorize the items one by one below.

Otherwise if we want to keep this "optimization", we could conditionally authorize the permission and then in case we receive a CONDITIONAL response, we manually check whether the response contains a `IS_TASK_OWNER` rule and in that case we perform authorization manually. But this makes things more tricky, so I'm not sure this is worth.

**Review by vinzscam — CHANGES_REQUESTED** [review]

File: plugins/scaffolder-backend/src/service/router.ts
Comment (Zaperex) [reply plugins/scaffolder-backend/src/service/router.ts:635]: Ah ok that makes alot of sense, I'll make the changes.

**Comment by acierto** [comment]
Hey @vinzscam 👋 

What would be your suggestion how to move forward?

**Comment by vinzscam** [comment]
> Hey @vinzscam 👋
> 
> What would be your suggestion how to move forward?

as in my comments we need to make sure the new rule definition is more generic, so something like `PermissionRule<TaskSpec, ...>` (or similar) instead of `PermissionRule<string, ...>` so that we won't have problem in the future to introduce more rules.

**Comment by drodil** [comment]
Hi @Zaperex , great work on this! Can you take a look at #27421 as we have some additional requirements. Of course the additional conditions can be added later but just wondering if they make sense and are doable?

**Comment by schultzp2020** [comment]
> Hi @Zaperex , great work on this! Can you take a look at #27421 as we have some additional requirements. Of course the additional conditions can be added later but just wondering if they make sense and are doable?

@vinzscam, would it be possible to merge this as is? @Zaperex has gone back to school and likely won’t be available for some time. Alternatively, is there anything I can do to help move this forward?

**Comment by CptnFizzbin** [comment]
> So on the whole I think that this is OK, although after chatting with some folks internally I think that there's a chance that this could be a breaking change, but only for the `alpha` exports as this is permissions, which marked as `alpha` for these exact reasons.
> ...

Would it be wrong to say that calling `isResourcePermission` without a resourceType is a bad practice?

**Pull request closed without merging**
